wandb/sdk/internal/writer.py

```diff
@@ -1,44 +1,200 @@
 
+    _ds: Optional[datastore.DataStore]
+    _flow_control: Optional[flow_control.FlowControl]
+    _status_report: Optional["pb.StatusReportRequest"]
+    _record_num: int
+    _telemetry_obj: tpb.TelemetryRecord
+    _telemetry_overflow: bool
+
+    # TODO(cancel_paused): implement me
+    # _sender_cancel_set: Set[str]
+
     def __init__(
         self,
-        settings,
-        record_q,
-        result_q,
+        settings: SettingsStatic,
+        record_q: "Queue[pb.Record]",
+        result_q: "Queue[pb.Result]",
+        sender_q: "Queue[pb.Record]",
+        interface: InterfaceQueue,
         context_keeper: context.ContextKeeper,
     ):
         self._settings = settings
         self._record_q = record_q
         self._result_q = result_q
+        self._sender_q = sender_q
+        self._interface = interface
         self._context_keeper = context_keeper
+
+        # TODO(cancel_paused): implement me
+        # self._sender_cancel_set = set()
+
         self._ds = None
+        self._flow_control = None
+        self._status_report = None
+        self._record_num = 0
+        self._telemetry_obj = tpb.TelemetryRecord()
+        self._telemetry_overflow = False
 
-    def open(self):
+    def open(self) -> None:
         self._ds = datastore.DataStore()
         self._ds.open_for_write(self._settings.sync_file)
+        self._flow_control = flow_control.FlowControl(
+            settings=self._settings,
+            write_record=self._write_record,
+            forward_record=self._forward_record,
+            pause_marker=self._pause_marker,
+            recover_records=self._recover_records,
+        )
+
+    def _forward_record(self, record: "pb.Record") -> None:
+        self._context_keeper.add_from_record(record)
+        tracelog.log_message_queue(record, self._sender_q)
+        self._sender_q.put(record)
+
+    def _send_mark(self) -> None:
+        sender_mark = pb.SenderMarkRequest()
+        record = self._interface._make_request(sender_mark=sender_mark)
+        self._forward_record(record)
+
+    def _maybe_update_telemetry(self) -> None:
+        if self._telemetry_overflow:
+            return
+        self._telemetry_overflow = True
+        with telemetry.context(obj=self._telemetry_obj) as tel:
+            tel.feature.flow_control_overflow = True
+        telemetry_record = pb.TelemetryRecordRequest(telemetry=self._telemetry_obj)
+        record = self._interface._make_request(telemetry_record=telemetry_record)
+        self._forward_record(record)
+
+    def _pause_marker(self) -> None:
+        self._maybe_update_telemetry()
+        self._send_mark()
+
+    def _write_record(self, record: "pb.Record") -> int:
+        assert self._ds
 
-    def write(self, record):
+        self._record_num += 1
+        proto_util._assign_record_num(record, self._record_num)
+        ret = self._ds.write(record)
+        assert ret is not None
+
+        (_start_offset, end_offset, flush_offset) = ret
```

Comment:
nit:
```suggestion
        _, end_offset, _ = ret
```

Reply:
I changed it to:
```
     _start_offset, end_offset, _flush_offset = ret
```
which i like better if flake8 is ok with it.